cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/WorkflowDriver.java

```diff
@@ -104,10 +111,42 @@ protected void run() throws Exception {
     ClassLoader classLoader = program.getClassLoader();
 
     // Executes actions step by step. Individually invoke the init()->run()->destroy() sequence.
-    Iterator<WorkflowActionSpecification> iterator = workflowSpec.getActions().iterator();
+
+    ApplicationSpecification appSpec = program.getSpecification();
+
+    Iterator<WorkflowActionEntry> iterator = workflowSpec.getActions().iterator();
     int step = 0;
     while (running && iterator.hasNext()) {
-      WorkflowActionSpecification actionSpec = iterator.next();
+      WorkflowActionSpecification actionSpec;
+      WorkflowActionEntry actionEntry = iterator.next();
+      switch (actionEntry.getType()) {
+        case MAPREDUCE:
+          MapReduceSpecification mapReduceSpec = appSpec.getMapReduce().get(actionEntry.getName());
+          if (mapReduceSpec == null) {
+            LOG.error("MapReduce program '{}' is not configured with the application.", actionEntry.getName());
+            throw new IllegalStateException("Workflow stopped without executing all tasks");
+          }
+          actionSpec = new DefaultWorkflowActionSpecification(new ProgramWorkflowAction(mapReduceSpec.getName(),
+                                                                                        mapReduceSpec.getName(),
+                                                 WorkflowSupportedProgram.MAPREDUCE));
+          break;
+        case SPARK:
+          SparkSpecification sparkSpec = appSpec.getSpark().get(actionEntry.getName());
+          if (sparkSpec == null) {
+            LOG.error("Spark program '{}' is not configured with the application.", actionEntry.getName());
+            throw new IllegalStateException("Workflow stopped without executing all tasks");
+          }
+          actionSpec = new DefaultWorkflowActionSpecification(new ProgramWorkflowAction(sparkSpec.getName(),
+                                                                                        sparkSpec.getName(),
+                                                 WorkflowSupportedProgram.SPARK));
+          break;
+        case CUSTOM_ACTION:
+          actionSpec = workflowSpec.getCustomActionMap().get(actionEntry.getName());
+          break;
+        default:
```

Comment:
Is this a test of the name or should there be a test of the type? (ie, actionEntry.getType() is what is unknown, rather than the name.)